source/extensions/filters/http/grpc_web/grpc_web_filter.cc

```diff
@@ -7,16 +7,54 @@
 #include "common/common/assert.h"
 #include "common/common/base64.h"
 #include "common/common/empty_string.h"
+#include "common/common/enum_to_int.h"
 #include "common/common/utility.h"
+#include "common/grpc/common.h"
 #include "common/grpc/context_impl.h"
 #include "common/http/headers.h"
 #include "common/http/utility.h"
+#include "common/runtime/runtime_features.h"
 
 namespace Envoy {
 namespace Extensions {
 namespace HttpFilters {
 namespace GrpcWeb {
 
+namespace {
+
+// This is arbitrarily chosen.
+// TODO(dio): Make it configurable when it is needed.
+constexpr uint64_t MAX_GRPC_MESSAGE_LENGTH = 16384;
+
+// This builds grpc-message header value from encoding buffer and the last frame (when
+// end_stream=true). When we have buffered data in encoding buffer, we limit the length of
+// grpc-message to be smaller than MAX_GRPC_MESSAGE_LENGTH. However, when we only have "last" data,
+// we send it all.
+std::string buildGrpcMessage(const Buffer::Instance* buffered, Buffer::Instance* last) {
```

Comment:
Again I don't think we can enforce limits here, if we want to handle local replies correctly

We can enforce limits if there are multiple encodeData calls.  Each encodeData call represents data we're already willing to buffer, and we want to make sure we handle local replies.

I'd suggest cribbing one of the tests from local_reply_integration_test to make sure we handle large local replies correctly.

Reply:
Added.

Reply:
I'm confused how you could have an integration test with local reply greater than max grpc messagae length and not trip the asserts below:
 ASSERT(buffered->length() <= MAX_GRPC_MESSAGE_LENGTH);

can you clue me in?

Reply:
From my understanding, `buffered` here is the encoding buffer, which is nullptr in the case of local reply.

Reply:
I have updated it to have two separate functions.